File: backend/app/modules/safety/schemas/fire_alarm.py
"""消防报警 Schema — API 契约（ticket 02 同步；ticket 03 记录查询/统计；ticket 05 日报/周报生成）。"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


class FireAlarmSyncResponse(BaseModel):
    """Bitable 全量同步结果。"""

    synced_count: int = Field(..., description="本次同步 upsert 的记录数")
    soft_deleted_count: int = Field(0, description="本次同步软删除的记录数")


class FireAlarmRecordOut(BaseModel):
    """报警记录列表项（含 AI 分析列）。"""

    id: uuid.UUID
    feishu_record_id: str | None = None
    source: str
    alarm_time: datetime | None = None
    alarm_type: str | None = None
    department: str | None = None
    department_leader_name: str | None = None
    building: str | None = None
    location: str | None = None
    alarm_nature: str | None = None
    cause_category: str | None = None
    cause_description: str | None = None
    ai_dimension: str | None = None
    ai_reason_analysis: str | None = None
    ai_rectification_direction: str | None = None
    ai_analyzed_at: datetime | None = None
    synced_at: datetime | None = None
    created_at: datetime | None = None


class FireAlarmStatsOut(BaseModel):
    """KPI 卡数据（契约字段以 ticket 03 为准，分布为 dict 而非数组）。"""

    date: str  # 当日日期 ISO（target_date 或今天·北京时间）
    today_total: int = 0  # 今日报警数
    week_total: int = 0  # 本周（自然周 周一~周日·北京时间）报警数
    total_count: int = 0  # 全量报警数（非软删）
    week_ai_analyzed_count: int = 0  # 本周已 AI 分析的记录数
    nature_distribution: dict[str, int] = Field(default_factory=dict)  # 报警性质 → 数量
    type_distribution: dict[str, int] = Field(default_factory=dict)  # 报警类型 → 数量
    dimension_distribution: dict[str, int] = Field(default_factory=dict)  # AI 维度 → 数量
    department_distribution: dict[str, int] = Field(default_factory=dict)  # 部门 → 数量


class FireAlarmDailyReportRequest(BaseModel):
    """日报生成请求（ticket 05；API 端点在 ticket 07）。"""

    target_date: date | None = Field(None, description="目标日期，默认今天（北京时区）")


class FireAlarmMonthlyReportRequest(BaseModel):
    """月报生成请求（原周报请求，2026-09-22 改月报）。"""

    month_end: date | None = Field(
        None, description="月内任一日期（取其所在自然月），默认上一完整自然月",
    )


class FireAlarmReportResponse(BaseModel):
    """日报/月报生成结果（不落库，前端弹窗展示）。

    report_kind: "daily" / "monthly"
    target_date: 日报=当日；月报=月末日期
    month_start: 月报=1 日（日报为 None）
    analyzed: 成功 AI 分析的记录数（仅日报）
    push_results: [{"chat_id", "success", "message_id"?, "skipped"?, "reason"?, "error"?}]
    records_analyzed: 已 AI 回写的记录 id（仅日报）
    """

    report_kind: str  # "daily" / "monthly"
    target_date: date  # 日报：当日；月报：月末日期
    month_start: date | None = None  # 月报：1 日（日报为 None）
    total: int  # 涉及记录数
    analyzed: int = 0  # 成功 AI 分析的记录数（日报）
    markdown_report: str
    push_results: list[dict] = Field(default_factory=list)
    records_analyzed: list[uuid.UUID] = Field(default_factory=list)

File: frontend/src/types/safety/fire-alarm.ts
// ============ 消防报警分析（fire-alarm-analysis）Types ============
// 契约对齐：ticket 03（records/stats，分布为 dict）、ticket 07（报告 report_kind/analyzed）、
// ticket 08 实现要点（ai_dimension 为英文枚举 process/operation/equipment/other）

/** AI 二次分析维度（英文枚举，后端归一化后回写；前端经 AI_DIMENSION_CONFIG 映射中文） */
export type FireAlarmDimension = 'process' | 'operation' | 'equipment' | 'other'

/** 报警记录（列表/详情共用，来源 safety.fire_alarm_records） */
export interface FireAlarmRecord {
  id: string
  feishu_record_id?: string | null
  source?: string | null // bitable / manual（预留）
  alarm_time?: string | null // ISO 时间，前端 dayjs 格式化
  alarm_type?: string | null
  department?: string | null
  department_leader_name?: string | null
  building?: string | null
  location?: string | null // 报警部位
  alarm_nature?: string | null // 报警性质（误报/真实报警等，飞书自由文本）
  cause_category?: string | null // 原因分类
  cause_description?: string | null // 具体报警原因（人工填写，AI 二次理解对象）
  // AI 分析字段组（日报生成时回写）
  ai_dimension?: string | null // 取值见 FireAlarmDimension（未知值前端降级灰 Tag 原文显示）
  ai_reason_analysis?: string | null
  ai_rectification_direction?: string | null
  ai_analyzed_at?: string | null
  synced_at?: string | null
  created_at?: string | null
  // 不暴露 is_deleted：后端列表默认过滤软删除记录，前端不感知
}

/** 列表查询参数（日期 YYYY-MM-DD） */
export interface FireAlarmQueryParams {
  page?: number
  page_size?: number
  date_from?: string
  date_to?: string
  department?: string
  alarm_type?: string
  alarm_nature?: string
  ai_dimension?: string
  keyword?: string
}

/** KPI 统计（ticket 03 契约：分布均为 dict[str, int]，统计口径默认本周） */
export interface FireAlarmStats {
  date: string // 当日日期 ISO（target_date 或今天·北京时间）
  today_total: number // 今日报警数
  week_total: number // 本周（自然周 周一~周日·北京时间）报警数
  total_count: number // 全量报警数（非软删）
  week_ai_analyzed_count: number // 本周已 AI 分析的记录数
  nature_distribution: Record<string, number> // 报警性质 → 数量
  type_distribution: Record<string, number> // 报警类型 → 数量
  dimension_distribution: Record<string, number> // AI 维度 → 数量
  department_distribution: Record<string, number> // 部门 → 数量
}

/** Bitable 全量同步结果 */
export interface FireAlarmSyncResult {
  synced_count: number
  soft_deleted_count?: number
}

/** 日报生成请求（target_date 可选，默认当天·北京时区） */
export interface FireAlarmDailyReportRequest {
  target_date?: string
}

/** 周报生成请求（week_end 可选，默认本周日） */
export interface FireAlarmWeeklyReportRequest {
  week_end?: string
}

/** 推送结果项 */
export interface FireAlarmPushResult {
  chat_id: string
  success: boolean
  message_id?: string
  skipped?: boolean // 未配置推送目标时跳过
  reason?: string
  error?: string
}

/** 日报/周报生成响应（不落库，前端弹窗展示） */
export interface FireAlarmReportResponse {
  report_kind: 'daily' | 'weekly'
  target_date: string // 日报：当日；周报：周末日期
  week_start?: string | null // 周报：周一（日报为 null）
  total: number // 涉及记录数
  analyzed: number // 成功 AI 分析的记录数（降级时 < total）
  markdown_report: string
  push_results: FireAlarmPushResult[]
  records_analyzed?: string[] // 已 AI 回写的记录 id（仅日报）
}
